planner/core/rule_generate_column_substitute.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/types"
+)
+
+type gcSubstituter struct {
+}
+
+// ExprColumnMap is used to store all expressions of indexed generated columns in a table,
+// and map them to the generated columns,
+// thus we can substitute the expression in a query to an indexed generated column.
+type ExprColumnMap map[expression.Expression]*expression.Column
+
+func (gc *gcSubstituter) optimize(ctx context.Context, lp LogicalPlan) (LogicalPlan, error) {
```

Comment:
Could you add a comment for this optimizaiton? We'd better explain the purpose and give some examples here.